# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [324]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

In [325]:
path = os.getcwd()
print ("The current working directory is %s" % path)

The current working directory is /Users/owner/Documents/Springboard/Capstone_Projects/Capstone_Projects/Guided_Capstone


**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [326]:
directory = "data"
data_dir = os.path.join(path, directory)
df = pd.read_csv(os.path.join(data_dir,'step3_output.csv'),sep='\t')

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [327]:
dfo=df[['state']] # select object type columns
df = pd.concat([df.drop(dfo, axis=1), pd.get_dummies(dfo)], axis=1)

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we don't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [328]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend','clusters'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend'] 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [331]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()


# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [332]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [333]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [334]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error

In [335]:
ev = explained_variance_score(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)

In [336]:
print("explained_vaiance_score: ",ev)
print("mean_absolute_error: ",mae)

explained_vaiance_score:  -1.2494839639380701e+23
mean_absolute_error:  936387891002.042


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [337]:
print(lm.intercept_)

-235404777008.68512


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [338]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
df_lm = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

In [339]:
print(df_lm.sort_values('Coefficient', ascending=False))

                       Coefficient
total_chairs          2.282087e+13
state_New York        1.355490e+13
surface               1.226970e+13
state_Michigan        1.167147e+13
double                1.160688e+13
state_New Hampshire   1.093160e+13
state_Wisconsin       1.053433e+13
triple                9.973568e+12
state_Pennsylvania    9.673894e+12
state_Minnesota       9.204574e+12
state_Vermont         8.165283e+12
state_California      8.165283e+12
state_Massachusetts   7.581527e+12
state_Idaho           7.581527e+12
state_New Mexico      7.581527e+12
state_Connecticut     6.940929e+12
state_Maine           6.940929e+12
state_Colorado        6.940929e+12
state_Wyoming         6.940929e+12
fastQuads             6.491311e+12
quad                  6.368317e+12
state_North Carolina  6.225970e+12
state_Ohio            6.225970e+12
state_Utah            6.225970e+12
state_Virginia        5.407231e+12
state_West Virginia   5.407231e+12
state_Illinois        5.407231e+12
state_Iowa          

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [340]:
# drop state column from dataframe
df2 = pd.read_csv(os.path.join(data_dir,'step3_output.csv'),sep='\t')
X2 = df2.drop(['state','Name','AdultWeekend','clusters'], axis=1)

In [341]:
# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scalar2 = preprocessing.StandardScaler().fit(X2)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X2_scaled=scalar2.transform(X2) 

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X2_train, X2_test, y2_train, y2_test = train_test_split(X2_scaled, y, test_size=0.25, random_state=1)

In [342]:
# modeling
lm2 = linear_model.LinearRegression()
model2 = lm2.fit(X2_train,y2_train)

In [343]:
# Make a variable called y2_pred and assign it the result of calling predict() on our model variable with parameter X_test
y2_pred = model2.predict(X2_test)

In [344]:
ev2 = explained_variance_score(y2_test, y2_pred)
mae2 = mean_absolute_error(y2_test, y2_pred)

print("explained_vaiance_score: ",ev2)
print("mean_absolute_error: ",mae2)

explained_vaiance_score:  0.775151728680886
mean_absolute_error:  6.512706378251336


In [345]:
print(lm2.intercept_)

57.51597976606887


In [346]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
df2_lm = pd.DataFrame(abs(lm2.coef_), X2.columns, columns=['Coefficient'])

In [347]:
print(df2_lm.sort_values('Coefficient', ascending=False))

                   Coefficient
AdultWeekday         10.804610
vertical_drop         3.746443
SkiableTerrain_ac     2.006130
summit_elev           1.738061
quad                  1.555589
daysOpenLastYear      1.386079
projectedDaysOpen     1.186394
fastSixes             1.091920
triple                0.935262
total_chairs          0.931033
Snow Making_ac        0.856981
surface               0.776056
fastQuads             0.652643
averageSnowfall       0.650662
NightSkiing_ac        0.579488
trams                 0.547855
TerrainParks          0.497322
double                0.466450
LongestRun_mi         0.408692
yearsOpen             0.165923
Runs                  0.163628


**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [348]:
# drop state column from dataframe
df3 = pd.read_csv(os.path.join(data_dir,'step3_output.csv'),sep='\t')
X3 = df3.drop(['state','Name','AdultWeekend','summit_elev','clusters'], axis=1)

In [349]:
# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scalar3 = preprocessing.StandardScaler().fit(X3)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X3_scaled=scalar3.transform(X3) 

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X3_train, X3_test, y3_train, y3_test = train_test_split(X3_scaled, y, test_size=0.25, random_state=1)

In [350]:
# modeling
lm3 = linear_model.LinearRegression()
model3 = lm3.fit(X3_train,y3_train)

In [351]:
# Make a variable called y2_pred and assign it the result of calling predict() on our model variable with parameter X_test
y3_pred = model3.predict(X3_test)

In [352]:
ev3 = explained_variance_score(y3_test, y3_pred)
mae3 = mean_absolute_error(y3_test, y3_pred)

print("explained_vaiance_score: ",ev3)
print("mean_absolute_error: ",mae3)

explained_vaiance_score:  0.7777668686037429
mean_absolute_error:  6.355600596679503


In [353]:
print(lm3.intercept_)

57.55586183157726


In [354]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
df3_lm = pd.DataFrame(abs(lm3.coef_), X3.columns, columns=['Coefficient'])

In [355]:
print(df3_lm.sort_values('Coefficient', ascending=False))

                   Coefficient
AdultWeekday         10.836588
vertical_drop         3.406817
SkiableTerrain_ac     2.557809
quad                  1.346843
averageSnowfall       1.245942
daysOpenLastYear      1.102341
total_chairs          0.990950
projectedDaysOpen     0.950726
fastSixes             0.932890
triple                0.882725
Snow Making_ac        0.873166
NightSkiing_ac        0.843134
double                0.620038
LongestRun_mi         0.598170
surface               0.594750
fastQuads             0.575337
trams                 0.476168
TerrainParks          0.394154
yearsOpen             0.102746
Runs                  0.052916


## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -1.2494839639380701e+23 |936387891002.042  ||
| Model 2. | 0.775151728680886|6.512706378251336 |'state'|
| Model 3. | 0.7777668686037429 |6.355600596679503 |'state','summit_elev','base_elev'|

Model Selection:

In [0]:
model3